### DISCUSSION TITLE
How to run ComputeCrystalPlasticityStress and get the stress-strain curve

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#17531] [Q&A Modules: Phase field] [answered]
Hi Moose expert,

I am new to Moose and looking through materials module and hope to use ComputeCrystalPlasticityStress class to calculate the response of a FCC crystalline solid. The example file given by Moose is https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.html.

The question is how I can run this example file to get the stress-strain or strain-time curve. I found there are many input files at the end of the page and have no idea how to compile and run them. This example seems to be a little more complicated example than those introduced in the get started for new users.

Thank you for your help in advance.

**Comment by PengWei97** [comment, marked as answer]
Hi xchenggood,
Add the following code to the input file, and make sure that the Outputs module `csv = true,` then you will find a csv file exists in this path, where Cauchy stress, Lagrangian strain and time exist.

> [Postprocessors]
>   [./stress_yy]
>     type = ElementAverageValue
>     variable = stress_yy
>     block = 'ANY_BLOCK_ID 0'
>   [../]
>   [./e_yy]
>     type = ElementAverageValue
>     variable = e_yy
>     block = 'ANY_BLOCK_ID 0'
>   [../]
> []
Finally, the specific input file is as follows,
[prop_grain_read_cp_03.i](https://github.com/PengWei97/monkey/blob/main/problem/poly_cp/prop_grain_read_cp_03.i)

**Reply by xchengood** [reply]
Really thank you for your guidance. It is much helpful. I look through the file you provide and download the entire folder 'monkey-main' on the up-level containing this file. When I try to use 'make -j8' that I learned in the get started for new users to compile prop_grain_read_cp_03.i, I got some errors. Could you tell me a little more about where I should put the folder 'monkey-main' to and how to compile and run prop_grain_read_cp_03.i correctly? Those examples in tutorial section have been put to the right path or directory. But I have no idea how to add this file to the right part and test it. Best regards.

**Reply by PengWei97** [reply]
> Really thank you for your guidance. It is much helpful. I look through the file you provide and download the entire folder 'monkey-main' on the up-level containing this file. When I try to use 'make -j8' that I learned in the get started for new users to compile prop_grain_read_cp_03.i, I got some errors. Could you tell me a little more about where I should put the folder 'monkey-main' to and how to compile and run prop_grain_read_cp_03.i correctly? Those examples in tutorial section have been put to the right path or directory. But I have no idea how to add this file to the right part and test it. Best regards.

hi @xchengood,
If there is no problem with my understanding, you can take a look at [Step 2: Write an Input File](https://mooseframework.inl.gov/getting_started/examples_and_tutorials/tutorial01_app_development/step02_input_file.html#step-2-write-an-input-file) to how to execute an input file. And if the execution is not successful, please send me the information output to the terminal after running to let me know where your problem lies.
wei peng

**Reply by xchengood** [reply]
@PengWei97 Thanks for your reply again. I try to execute prop_grain_read_cp_03.i using commands introduced in the Step2 and I got the following error. How can I fix it? Thank you.

![image](https://user-images.githubusercontent.com/74226455/114199515-fd6ddd80-9919-11eb-9faf-5e54b530eded.png)
Best wishes!

**Reply by jiangwen84** [reply]
what is the executable file you are using? 

You should compile the moose by running `make -j 8`  under `/moose/modules/` for combined applications  or `/moose/modules/tensor_mechanics/` for mechanics applications.

**Reply by ngrilli** [reply]
Dear @xchengood 

The way I usually do is to create an application as described in:
https://mooseframework.inl.gov/getting_started/new_users.html#create-an-app

Make sure you enable all the modules in the Makefile before compiling.
The executable created should run your input file properly.

The ElementPropertyReadFile is part of the tensor mechanics module,
therefore, as @jiangwen84 suggests, your executable is not linking properly with that specific module.

Best Regards,
Nicolò Grilli

**Reply by PengWei97** [reply]
hi @xchengood,
As suggested by @jiangwen84 and @ngrilli, you should modify `the Makefile file` after creating your own application, make sure `TENSOR_MECHANICS := yes`, then `make -j4` to compile them once to generate the *-opt file, and finally type in the terminal `*-opt -i prop_grain_read_cp_03.i` to execute the input file of moose.

Best Regards
Wei

**Reply by xchengood** [reply]
@jiangwen84 @ngrilli @PengWei97. Really thank you for your helpful suggestions.  I succeed in getting output with csv format and the curve is drawn below. Based on my limited understanding, the curve seems to be weird and I am confused about it shows a linear relationship.
![image](https://user-images.githubusercontent.com/74226455/114424421-b5e78b80-9b7d-11eb-9ba4-4610c2db2941.png)
![image](https://user-images.githubusercontent.com/74226455/114422568-0362f900-9b7c-11eb-855f-7d8ef05c9af5.png)
Based on this curve, there are some bemused points in my head.
1. This output seems the stress as a function of time. How can I get strain-time or strain-stress curve? what is unit for time?
2. How can I check and visualize the generated mesh?
3. About material definition, how can I define my own material property?

I appreciate your help.

**Reply by PengWei97** [reply]
hi @xchengood,
Can you display the input file you are writing now? In this way, I can give you some preliminary suggestions for your three questions. After all, I am also a moose rookie.

Best Regards
Wei

**Reply by xchengood** [reply]
Hi @PengWei97. The input file I used is the same as the one you mentioned in https://github.com/PengWei97/monkey/blob/main/problem/poly_cp/prop_grain_read_cp_03.i. The 'input_file.txt' in programming is also the same as the one in the poly_cp folder. Thanks.

**Reply by PengWei97** [reply]
hi @xchengood,

>  The 'input_file.txt' in programming is also the same as the one in the poly_cp folder.

You should input the input_file.txt file with 200 sets of Euler angles, like this,  [input_file.txt](https://github.com/idaholab/moose/files/6316215/input_file.txt), otherwise the simulation will report an error. And by reading ElementPropertyReadFile.C, you will find that when nprop = 3, it is not necessary to use the last column of the input_file.txt file provided by moose.

> 1. This output seems the stress as a function of time. How can I get strain-time or strain-stress curve? what is unit for time?

In this way, using the Euler angle input file I just provided, simulate the uniaxial stretching of 200 grains, and finally get a file named [prop_grain_read_out_xiu03.csv](https://github.com/PengWei97/monkey/blob/main/problem/poly_cp/prop_grain_read_out_xiu03.csv), which contains time(s), Lagrangian strain, and Cauchy stress. And you can output the stress-strain curve, as shown in the figure,
![image](https://user-images.githubusercontent.com/56183775/114838021-e130db80-9e06-11eb-993a-a4f617f03149.png)
After that, the unit system of the crystal plasticity module is mm-MPa-s. Please forgive me that I forgot the specific moose URL.

> 2. How can I check and visualize the generated mesh?

First, moose has its own visualization software, peacock. Second, you can use paraview, a software to visualize the mesh you generate, which is also introduced on the moose website.

> 3. About material definition, how can I define my own material property?

Defining your own material properties may be more difficult than you think. You may need to refer to the existing material modules of moose and its introduction to writing C++ code in the introductory files. Please forgive me that I am currently trying to continue the modification of the material module, and I have no more experience to share with you. hi @jiangwen84 and @ngrilli, Would you have any good experiences to share in this regard?

Best Regards
Wei

**Reply by xchengood** [reply]
Hi @PengWei97 . Thanks a lot. Due to my limited understanding, I originally thought e_yy as a calculation error accuracy. Now, I am clear it is the Lagrangian strain (mm). The curve you gave does make sense.

About the question about how I can check and visualize the generated mesh, as you mentioned, the peacock is a good tool to visualize the generated mesh. But there already is a mesh file with .e format in the example section and it can be checked via peacock directly. But usually, the mesh generation is controlled and achieved by the following similar codes. After running, I don't see a mesh file in the output. So how can I visualize it?
![image](https://user-images.githubusercontent.com/74226455/114894084-ed4f7580-9dd3-11eb-947a-22ced6e34765.png)

In addition, could you share good experiences and example input files about the simulation of the thermal creep behaviour of materials with me? It involves a constitutive model, as mentioned in this report. https://link.springer.com/article/10.1007/s11661-017-4011-3.

Thank you.

**Reply by PengWei97** [reply]
hi @xchengood,

Perhaps you should set `exodus = true` in your input file `Outputs`, so that you can use the two software to visualize the mesh files generated by your simulation.
Regarding your second question, I'm sorry I can't do anything about it. Perhaps the first two experts can give you some guidance.

wei

**Reply by xchengood** [reply]
> hi @xchengood,
> 
> Perhaps you should set `exodus = true` in your input file `Outputs`, so that you can use the two software to visualize the mesh files generated by your simulation.
> Regarding your second question, I'm sorry I can't do anything about it. Perhaps the first two experts can give you some guidance.
> 
> wei

Really thank you. I learned a lot from your response. Let me study moose more and hope to acquire it soon.

**Comment by THOMPSONFLORIDA** [comment]
Hello all
i am working on adding the impact of elastic stress on a corrosion model using phase field(MOOSE) 
i need your contribution on how to fomulate the free energy expression

or share an example input file

**Reply by GiudGiud** [reply]
Hello

The phase field module has several pages on how to formulate free energies. You could start from there
https://mooseframework.inl.gov/modules/phase_field/index.html

Then once you are comfortable with that and with thermo-mechanics, you should check out some examples of coupling in the combined module
https://github.com/idaholab/moose/tree/next/modules/combined/examples/phase_field-mechanics

PS: next time, open your own post please ;)

Guillaume
